blockly-rtc/src/MarkerManager.js

```diff
@@ -0,0 +1,149 @@
+import MarkerUpdate from './MarkerUpdate';
+
+export default class MarkerManager {
+  constructor(workspaceId, sendMarkerUpdate, getMarkerUpdates,
+      addClient, getBroadcastMarkerUpdates) {
+    this.workspaceId = workspaceId
+    this.colours =  [
+        '#fcba03', '#03fc20', '#03f0fc', '#035efc', '#5603fc', '#fc03d2'];
+    this.sendMarkerUpdate = sendMarkerUpdate;
+    this.getMarkerUpdates = getMarkerUpdates;
+    this.addClient = addClient;
+    this.broadcastMarkerHandler = getBroadcastMarkerUpdates;
+  };
+
+  /**
+   * Initialize the workspace by creating and registering markers for all active
+   * users and activating the handling of recieving MarkerUpdates from the server.
+   * @public
+   */
+  async initMarkers() {
+    await this.addClient(this.workspaceId);
+    const markerUpdates = await this.getMarkerUpdates();
+    markerUpdates.forEach((markerUpdate) => {
+      this.createMarker_(markerUpdate);
+      if (this.broadcastMarkerHandler) {
+        this.broadcastMarkerHandler(this.updateMarkerLocations_.bind(this));
+      } else {
+        this.pollServer_();
+      };
+    });
+  };
+
+  /**
+   * Create a MarkerUpdate from a Blockly event and send it to the server.
+   * @param {!Blockly.Event} The event for which to create a MarkerUpdate.
+   * @public
+   */
+  async handleEvent(event) {
+    const markerUpdate = MarkerUpdate.fromEvent(event);
+    await this.sendMarkerUpdate(markerUpdate);
+  };
+
+  /**
+   * Periodically query the database for MarkerUpdates.
+   * @private
+   */
+  async pollServer_() {
+    const markerUpdates = await this.getMarkerUpdates();
+    await this.updateMarkerLocations_(markerUpdates);
+    setTimeout(() => {
+      this.pollServer_();
+    }, 5000)
+  };
+
+  /**
+   * Get the workspace with the id specified by the MarkerUpdate.
+   * @return {Blockly.Workspace} The sought after workspace or null if not found.
+   * @private
+   */  
+  getWorkspace_() {
+    return Blockly.Workspace.getById(this.workspaceId);
+  };
+
+  /**
+   * Get the MarkerManager for the workspace specified by MarkerUpdate.
+   * @return {Blockly.MarkerManager} The sought after MarkerManager or null if not
+   * found.
+   * @private
+   */  
+  getMarkerManager_() {
+    return this.getWorkspace_().getMarkerManager();
+  };
+
+  /**
+   * Get a unique color to assign to a new client marker.
+   * @returns {string} The string encoding a colour.
+   * @private
+   */  
+  setColour_() {
+    return this.colours.pop();
+  };
+
+  /**
+   * Create a Marker with a unique color and register it.
+   * @param {!MarkerUpdate} markerUpdate The MarkerUpdate for which to create
+   * a Marker.
+   * @returns {!Marker} marker The Marker represented by the MarkerUpdate.
+   * @private
+   */
+  createMarker_(markerUpdate) {
+    const marker = markerUpdate.toMarker(this.getWorkspace_());
+    marker.colour = this.setColour_();
+    this.getMarkerManager_().registerMarker(markerUpdate.id, marker)
```

Comment:
`getMarkerManager_` can return null.

Reply:
Added a check.

Reply:
I think this issue still exists.

Reply:
I now changed it so that it throws an error if MarkerManager returns null and added a test for this.